### DISCUSSION TITLE
Version 2

### DISCUSSION THREAD

**Discussion by leaanthony** [wailsapp/wails#662] [General]
This is for discussing anything related to Version 2 of the project.

**Comment by SGarno** [comment]
I noticed your posting regarding Version 2  in that there are plans to support WebView2.    

Where are you with that in the 2.0.65 alpha?

**Reply by leaanthony** [reply]
Absolutely! The `v2-alpha` branch already has support but not all of the Wails API is currently ported. Be warned though, the windows stuff is quite unstable right now. That being said it's fundamentally working:
![image](https://user-images.githubusercontent.com/1943904/118197851-1d516d80-b493-11eb-88d6-0422872a47bb.png)

**Comment by SGarno** [comment]
Awesome!  Well, I would be willing to give it a go and provide feedback if that helps.

**Comment by SGarno** [comment]
I have fixed a few things with the windows side of the house on the build and vscode debug.    I could do a pull request if that would be helpful.

**Reply by leaanthony** [reply]
Perfect! I was going to tackle winres next so if it's related to that then awesome 😃

**Reply by SGarno** [reply]
I am not that deep into Wails2 yet.   What is the winres issue?

All I fixed was the pathing on wails init for windows.   VSCode understands / as well as \ in file paths so I changed it to use forward slash in the vscode files.   Filepath.join on windows returns a single backslash and when that is replaced into the templates, it becomes `\windows` instead of `\\windows`.   Since you had a hardcoded / in the template, I updated the filepath.join to use filepath.ToSlash() so that it was consistent.

Also, there is an inconsistency in the pathing structure in the build vs debugging in vscode.    When you run build, it goes to build/bin, but in the vscode tasks.json it is build/windows/server and build/windows/desktop.    What is the go-forward path that wails will use?

Lastly, if I make changes, do you want one pull request per change or batch up a bunch of commits in one pull request?

**Reply by leaanthony** [reply]
I've slightly changed the build directory and what goes in there so you may want to pull latest. Basically moved `assets` to `build` and all binaries go to `build/bin` by default. Windows will be `name.exe`, linux/mac will be `name-GOOS-GOARCH` by default. For non desktop builds, they will have `-BUILDTYPE` in the filename part.

Any fixes welcome and thanks for trying it out!

**Reply by SGarno** [reply]
Ok, so I have made a few fixes and am about to create a pull request.   Here are some of the changes I made:

- Fixed an issue with hardcoded / in a filename used for assets.   It now uses os.PathSeparator as the delimiter.   I found this was the real issue rather than using ToSlash().
- Fixed a problem with the scanning of the html file for asset inclusion.   There were extra tokenizer.Next() calls that was skipping some of the assets.
- Fixed the debug build on windows.   Windows requires the .exe extension, so I added it to the debug build.  

What I am struggling with at the moment is getting the build to recognize assets in a subdirectory.   I can see that it is including them during AssetLoad, but when you go to use them in the HTML, it doesn't seem to know they are there.    Am I missing something in the config or build process?

In the HTML, (after the webpack) I have 

```<body>
    <h1>Testing</h1>

    <p style="border-style: double;">
        <img src="img/68ae1645.svg">
    </p>
    <div id=xq-app>
    </div>
    <script src=js/vendor.30ef1ab5.js></script>
    <script src=js/app.e6a219ce.js></script>
</body>
```

I can open this up from the file explorer and it works fine, but in the built wails app, no image shows up.

I can't trace anything because the web debugger quits on refresh, but from what I can tell the assets are not there.   It gives me chunk error.  

Ideas?

**Reply by leaanthony** [reply]
How are you bundling the image? Are you using a plugin? I use base64 images and there's plugins to do that automatically

**Reply by SGarno** [reply]
What I am doing is trying to make Wails work with other ui builders native CLIs such as Vue, Nuxt, Quasar, Veutify, etc.    I am currently using [Quasar v2 beta](https://quasar.dev) as my starting point.    Here is what I did:

1. Initialize wails project `wails init -d sample2 -g -n sample2 -vscode`
2. Delete the frontend directory that was created `cd sample2 && rmdir /s frontend`
3. Create the quasar app into the frontend directory `quasar create`
4. Update the wails.json with the following:
```
              "html": "frontend\\dist\\spa\\index.html",
              "frontend:build" : "quasar build"
```
5. Update main.go for windows adding in the lines:
```
 		Windows: &windows.Options{
			WebviewIsTransparent:          true,
 			WindowBackgroundIsTranslucent: true,
 		},
```
6. Run the build `wails build`

The `quasar build` command compiles and webpacks the code into the `frontend/dist/spa` directory which I am pointing wails to.   At this point it should be nothing more than assets being included.   I have traced wails (this is where I found the tokenizer bug) and I can see that it is going through the links and img tags and incorporating them into the embedding.    Just not sure after that, why it can't seem to find the assets because the edge debug window disappears when I try to put breakpoints in and trace it.

The ultimate goal here is to allow wails to operate with other frontend cli's using their native capability (as opposed to requiring a specialized template) if possible. 

I would also like to get `quasar dev` during `wails dev` using bridge mode (I haven't gotten to that point yet).   

(edit)
In case it helps....here is the exact error in the console:
```
Pushing event listener: wails:sync:store:updatedbybackend:wails:isdarkmode
VM9:1 Pushing event listener: wails:sync:store:updatedbybackend:wails:loglevel
VM9:1 Pushing event listener: wails:sync:store:updatedbybackend:wails:appconfig
VM9:9 ChunkLoadError: Loading chunk 3 failed.
(error: )
    at Function.i.e (<anonymous>:9:81978)
    at component (<anonymous>:9:83875)
    at <anonymous>:9:16772
    at <anonymous>:9:17022
    at Array.map (<anonymous>)
    at <anonymous>:9:16998
    at Array.map (<anonymous>)
    at It (<anonymous>:9:16948)
    at <anonymous>:9:16458
    at p (<anonymous>:9:20183)
```

**Reply by leaanthony** [reply]
Are you trying to bundle images? This won't happen automatically and you'll need to use a bundler plugin to do it. Remember, there's a big difference between a site/app that has all its files on the local FS and a Wails app where everything is bundled into the binary. If you reference local paths, then don't be surprised if your app can't load them. What would be amazing is if we *could* bundle those kinds of assets without relying on a 3rd party plugin. We actually already have the start of that in the `html` package, but it isn't a priority as there are plenty of solutions for frontend bundling that already exist. 

One last thing - however you bundle your frontend, chunking is probably not what you want to do.

**Reply by SGarno** [reply]
Ok, I think I am confused then.    In v2/internal/html is assetbundle.go, the function loadAssets calls processHTML which seems to be adding the assets that are referenced in the html file.   Is that not the case?

**Reply by leaanthony** [reply]
Not entirely: it looks for the JS and CSS files from the HTML. The reason we've done this is that there is an experimental server mode that will run your app headless and that will deal with assets a little bit differently to desktop targets.

There's a case to be made for bundling images but what if they are remote images? Do you leave them be or pull them in? What if the image is being created dynamically? At the moment, we assume that all the image/font bundling is done by the developer via EG webpack and will "just work" by executing the javascript. If you want tips on how that works, look at the v1 templates.

**Comment by AlienRecall** [comment]
i'am waiting for version 2.0 since my app doesn't work on windows because some react libraries doesn't load on IE11. Is there any fix for that or i need to wait for the 2.0 release?

**Reply by leaanthony** [reply]
I can't fix ie11 or react libraries 😃 you could try the alpha and see if they work?

**Reply by AlienRecall** [reply]
Is there any template I can follow to write a go + react app with wails v2? Because I'm moving from v1 and I'm getting a problem when running `wails build`.
I was getting `ERROR: open /Users/alienrecall/go/src/goguiv2/frontend/build/https:/fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap: no such file or directory` then I've removed the font from index.html and then it successfully complied the application, but it was giving me a javascript error.

**Reply by AlienRecall** [reply]
i was trying in development mode and when running npm run start the chrome console was showing `Cannot read property '_' of undefined` in file init.js line 172. I think it's something releated with the Wails.Init function in index.html

**Reply by AlienRecall** [reply]
**UPDATE**
I've tried also with vitejs + react but nothing changed, still getting the same error
```
init.js:18 Uncaught TypeError: Cannot read property '_' of undefined
    at Object.Init (init.js:18)
    at main.jsx:31
```

**Reply by leaanthony** [reply]
There isn't much in the way of docs for V2 so this is probably why you're getting unstuck. Try this:

```js
import { ready } from '@wails/runtime'

ready( ()=> {
  // init your JS code
});
```

Example from [xbar](https://github.com/matryer/xbar):
```js
import App from './App.svelte'
import { ready } from '@wails/runtime'
import { routes } from 'svelte-hash-router'
import Homepage from './Homepage.svelte'
import PluginsList from './PluginsList.svelte'
import PluginView from './PluginView.svelte'
import PeopleView from './PersonView.svelte'
import InstalledPluginView from './InstalledPluginView.svelte'

let app;

routes.set({
	'/': Homepage,
	'/plugins/*': PluginsList,
	'/installed-plugins/*': InstalledPluginView,
	'/plugin-details/*': PluginView,
	'/people/:username': PeopleView,
})

ready(() => {
	app = new App({
		target: document.body,
	});
});

export default app;
```

**Reply by AlienRecall** [reply]
```node
import React from 'react';
import ReactDOM from 'react-dom';
import 'core-js/stable';
import './index.scss';
import App from './App';
import * as serviceWorker from './serviceWorker';
import "@fortawesome/fontawesome-free/css/all.min.css";
import { ready } from '@wailsapp/runtime';
import { I18nextProvider } from "react-i18next";
import i18next from "i18next";


import common_it from "./translations/it/common.json";
import common_en from "./translations/en/common.json";

i18next.init({
  interpolation: { escapeValue: false },  // React already does escaping
  lng: 'en',                              // language to use
  resources: {
    en: {
      common: common_en               // 'common' is our custom namespace
    },
    it: {
      common: common_it
    },
  },
});


ready(() => {
  ReactDOM.render(
    <React.StrictMode>
      <I18nextProvider i18n={i18next}>
        <App />
      </I18nextProvider>
    </React.StrictMode>,
    document.getElementById("app")
  );
});

serviceWorker.unregister();
```

this is my code, is it correct? Because now I'm getting `ready is not a function` so I think I'm probably doing something wrong

**Reply by leaanthony** [reply]
Looks like you are using `wailsapp/runtime` (v1) instead of `wails/runtime` (v2)

**Reply by AlienRecall** [reply]
thanks that solved the problem, but I've noticed that with react I'm getting `Cannot assign to read only property 'exports' of object '#<Object>'` (but I think it's an issue related with webpack), then with react + vite I get `Target container is not a DOM element.` at bridge.js line 1646

**Reply by AlienRecall** [reply]
never mind I've just noticed the issue https://github.com/wailsapp/wails/issues/685 and changing the import to require I've fixed the react native problem (but not the vite problem). I getting troubles using events, I'm trying to send and event from backend to the frontend, on the cli seems all good, the event is sent by backend correctly, but is not received by the frontend, the chrome console says:
```
 wails bridge  Unknown message: n{"name":"group:getAll","data":[[{"id":"572d0fa1-75f2-4b83-94d2-4138a0c0b83c","name":"NAME"}]]}
```
the cli says `TRACE | Event Subsystem | No listeners for event 'group:getAll'`

here is the frontend code:
```node
import { Events } from '@wails/runtime'

useEffect(() => {
        window.backend.managers.RuntimeManager.GetAllGroups(); //<- trigger the event
        Events.On('group:getAll', data => {
            handleGroupCreate(data)
        })
}, [handleGroupCreate])
```

**Reply by leaanthony** [reply]
Hmm.. that was addressed by this PR: https://github.com/wailsapp/wails/pull/700 - are you sure you have latest?

**Reply by AlienRecall** [reply]
yes, I have the latest

**Reply by AlienRecall** [reply]
it's strange because in the chrome console I get 
```
Pushing event listener: group:getAll
Pushing event listener: group:getAll
Pushing event listener: group:deleted
Pushing event listener: group:getAll
Pushing event listener: group:deleted
Pushing event listener: group:getAll
Pushing event listener: group:deleted
Pushing event listener: task:getTask
```
also multiple times, but then I get `No listeners for event 'group:getAll'` in the wails terminal

*this appens also on other events*

**Reply by leaanthony** [reply]
Is this a project that I can clone to test?

**Reply by AlienRecall** [reply]
where can I contact you to send you the project?

**Reply by leaanthony** [reply]
You can invite me if it's private or paste the link here if it's public.

**Reply by AlienRecall** [reply]
done, I've invited you in a demo repository, also feel free to use it as wails v2 + react template

**Reply by AlienRecall** [reply]
any news?

**Reply by leaanthony** [reply]
A couple of things to start:

  * You are including both v1 and v2 runtimes. This will not end well!
  * You seem to have ported v1 template instead of making a fresh frontend (i can see wails-react-scripts)

If I were you, I'd delete everything from frontend, generate a react template using whatever the current fashion dictates, then add the `ready()` hook into your project. Once you get a project working, you can convert it to a template.

**Comment by twiny** [comment]
Hello,
I've create a small template to quickly start a Wails v2.0.0-alpha.65, Svelte & Tailwindcss.
[github.com/superiss/desktop](https://github.com/superiss/desktop)

Application works perfectly on MacOS but I m wondering how to build my app to window or linux in "v2.0.0-alpha.65"?

**Reply by leaanthony** [reply]
Hey that's awesome! If you have a windows machine/VM you should be able to test it as basic templates are working on Windows right now. Haven't ported cross compilation yet. Failing that, make a PR and I can test. Linux isn't quite ready yet.

**Reply by twiny** [reply]
Got it! I will try! Thank you

**Reply by leaanthony** [reply]
Let us know if you have any issues as I've literally just compiled a working windows app:
![image](https://user-images.githubusercontent.com/1943904/118469198-d5f40700-b748-11eb-8818-7e2a5715b747.png)

**Reply by twiny** [reply]
Hey @leaanthony, sorry i did not had access to a windows machine till today, unfortunately i was not able to run my app on windows. i must be doing something wrong.

<img width="1431" alt="wails-dev" src="https://user-images.githubusercontent.com/59224605/118880478-d081f280-b8e9-11eb-9430-156e88a82f50.PNG">

On dev mode. wails finish compiling the app. but the .exe was not found in the dev/windows directory

<img width="785" alt="wails-build" src="https://user-images.githubusercontent.com/59224605/118880485-d2e44c80-b8e9-11eb-9792-c6edf942c5ac.PNG">

**Reply by leaanthony** [reply]
You'll need to clone the repo, switch to the `v2-alpha` branch, cd `v2/cmd/wails` and go install.

I'll try and get some docs up tonight. Be aware, there has been some slight breaking changes to project structure so you may have to update your local project (`asset` has been renamed to `build`).

**Reply by twiny** [reply]
@leaanthony Awesome. thank you very much. I will try it out tomorrow morning. i will post update.

**Reply by leaanthony** [reply]
Thanks for being an early tester! It's *really* hard to get everything done, including testing, so all help massively appreciated.

**Reply by twiny** [reply]
Hello,

I cloned the repo, and install the Wails v2 command from `v2-alpha` branch.

this time `wails build` finish building without errors. but when i click on the app .exe. the app does not open. however on `wails dev` mode, i got a connection error to the bridge.

<img width="693" alt="wails-dev-build-2" src="https://user-images.githubusercontent.com/59224605/119198653-99dee000-ba81-11eb-95a4-e6b06c10ebac.PNG">


backend debug:

<img width="1458" alt="wails-dev-build" src="https://user-images.githubusercontent.com/59224605/119198744-c09d1680-ba81-11eb-9166-8f0edb6393ee.png">

platform not support. from this: https://github.com/wailsapp/wails/blob/31cf04a9449deb0c0cbccb3522e7fad702dd13c2/v2/internal/bridge/session.go#L78

`wails doctore` output
<img width="620" alt="wails-doctor" src="https://user-images.githubusercontent.com/59224605/119198977-2c7f7f00-ba82-11eb-834a-e007705209b5.PNG">


On **MacOS** it works as expected. here is the example of the app on the test: [https://github.com/issmeftah/wails-template](https://github.com/issmeftah/wails-template)

I m not sure what i m missing.. installed the cmd from `v2-alpha` branch.

**Reply by twiny** [reply]
@leaanthony, i can maybe write some doc for `wails v2`, tell me where to start

**Reply by leaanthony** [reply]
Thanks for the offer! I see you've done great work with your template. The Windows port is currently underway so don't be surprised if it's not 100% working right now. I'll try and clone your template and debug it.

**Comment by svengreb** [comment]
First things first: This is a great project, thanks a lot for your effort 💪🏼
It is finally a solution to strip Electron from my projects. I used Node's IPC module and native Linux sockets (pipes on Windows) to communicate with a Go application that was embedded into the Electron bundle and executed as child process when Electron started. This way I was able to do all the "heavy work" in Go and let Electron only do the rendering with React by pushing messages via IPC. Wails on the other side is exactly was I was looking for all the time and I've spend some months now to create a usage concept for my projects and already started to port some of them (with absolute success 🎉).
Sorry for the small off-topic, but I know that feedback is important for OSS maintainers.

After porting one of my projects to Wails v1 some months ago I started to migrate it to v2, but got stuck when running the development mode. After spending some hours with (unsuccessfully) debugging my own code. I suspected that the chance it high that there is a problem in the Wails v2 alpha branch itself so I started to debug with a local repository clone of Wails. The main problems were two JS problems when running in development mode:

- `undefined is not a object (evaluate window.webkit.messageHandlers)`
- `undefined is not a object (evaluate main.Basic...)`

So basically somehow the `window.webkit` was not defined at all so calls to keys like `messageHandlers` and `messageHandlers.external.postMessage` also failed. The second error indicated that the Go <-> JS bindings are not available and checking the `window.backend` object only returned and empty object (since `{}` is set as default in the code).
After digging into the latest commits I noticed that the `setupIPCBridge` function of the `v2/internal/runtime/js/runtime/bridge.js` file has been changed and the definition of the `window.webkit` has been changed in [wailsapp/wails@d2e4a0b0][6] to use the `window.wailsInvoke` function instead which explains the first JS error.

```diff
function setupIPCBridge() {
	-// darwin
	-window.webkit = {
	-	messageHandlers: {
	-		external: {
	-			postMessage: (message) => {
	-				websocket.send(message);
	-			}
	-		},
	-		windowDrag: {
	-			postMessage: () => {
	-				// Ignore window drag events
	-			}
	-		}
	-	}
	+window.wailsInvoke = (message) => {
	+	websocket.send(message);
	+};
}
```

The error regarding the empty `window.backend` object is a result of the first error since the IPC message is never send and bindings are not set up.
Note that the problem is likely only relevant for macOS users since the function call was adjusted in the `windows.js` file.

To fiy both errors I've adjusted the `SendMessage` function in the [`v2/internal/runtime/js/desktop/darwin.js`][3] file to use `window.wailsInvoke` instead:

```diff
export function SendMessage(message) {
  -window.webkit.messageHandlers.external.postMessage(message);
  +window.wailsInvoke(message);
}
```

After rebuilding the assets (using the [`rebuild.go`][4] one-shot-runner) everything worked fine again. However, I'm not that familiar with the code to say that this is the only adjustments that is necessary. I found some more places where the `window.webkit.messageHandlers.external.postMessage` is called, i.e. in the [`v2/internal/ffenestri/ffenestri_darwin.c`][5] file.

Since these changes were published in [`@wails/runtime@v1.13.18][1], the development mode and (production) builds will break for all macOS users who did not pin the npm package to an explicit version. Using `@wails/runtime@<=v1.13.17` works fine.

**TL;DR;**

The npm package version [`@wails/runtime@v1.13.18`][1] breaks builds for macOS users due to the `darwin` JS runtime changes made in [`wailsapp/wails@d2e4a0b0`][2]. Adjusting the `SendMessage` function in the [`v2/internal/runtime/js/desktop/darwin.js`][3] file fixes the problem, but it might require more changes because it looks like there are other places where the `window.webkit.messageHandlers.external.postMessage` function is still called, i.e. in the [`v2/internal/ffenestri/ffenestri_darwin.c`][5] file.

Maybe you already noticed this bug on your local system already, but haven't pushed a new commit yet. Anyway, just wanted to share this finding to save other (and you) from spending more hours into finding the root cause.

[1]: https://www.npmjs.com/package/@wails/runtime/v/1.3.18
[2]: https://github.com/wailsapp/wails/commit/d2e4a0b0d203d7d7d5b1dd288cc73226089d940c#diff-371f1b929a8ed65c05ebdc846e6be495a185b3294434be4422153d4763737a85
[3]: https://github.com/wailsapp/wails/blob/4738a0e0a87bd4cf5afdd8d05a7e88bfd83e8570/v2/internal/runtime/js/desktop/darwin.js#L21-L23
[4]: https://github.com/wailsapp/wails/blob/0b8056bc/v2/internal/runtime/scripts/rebuild.go
[5]: https://github.com/wailsapp/wails/blob/1344638c5/v2/internal/ffenestri/ffenestri_darwin.c#L986
[6]: https://github.com/wailsapp/wails/commit/d2e4a0b0d203d7d7d5b1dd288cc73226089d940c#diff-371f1b929a8ed65c05ebdc846e6be495a185b3294434be4422153d4763737a85

**Reply by leaanthony** [reply]
Hi @svengreb! Thanks for the kind words about the project. It's been a lot of hard work so positive feedback is great!

Regarding dev mode, yes, I made some changes a couple of days ago to get the windows builds working in dev mode. To give you some context, the IPC Bridge was, as you could see, only taking Mac's webkit convention into consideration. Initially I added Window's equivalent to this method but realised this was messy. The answer was to wrap the comms in a single `wailsInvoke` method that, for bridge mode just sends that over websocket. This method is implemented by ffenestri for the desktop builds. This meant working back and removing all platform specific calls to the message handlers. For windows, this works great. I just hadn't got round to doing Mac yet. There should be no `messagehandlers.external.postMessage` anywhere in the runtime, just `wailsInvoke`.

Thanks for taking the time to dig into the code and I'm really glad you found the error. I'll push a fix for that in the next couple of hours!

**Reply by leaanthony** [reply]
@svengreb `@wails/runtime v1.3.19` released, darwin runtime updated, vanilla template updated. Both `wails build` and `wails dev` tested on both Mac and Windows.

Thanks for your detailed report and good luck!

**Reply by svengreb** [reply]
That was fast, thanks! 👍🏻

**Reply by svengreb** [reply]
Alright, after fighting with this unusable pixel mash called "Windows" through a VM, I finally got it working :smile:
Maybe I've missed something, but I guess since `v2` is still in development there are no detailed documentations yet except the [roadmap and project plan wiki page][1].
After trying to compile the application for `windows/amd64` within a container environment (Docker with [crazy-max/xgo][2]), I gave up and guess it is currently not possible due to the Windows specific dependencies and native WebView2 libraries. So instead of only using the Windows VM for testing I decided to also use it to compile the artifact for now. The building itself works fine, but when trying to run the application it either only rendered the window for a millisecond and exited again or (for whatever "lucky" reason) the window stayed open but only rendered as (white) blank box. Many hours and debugging tries later I had the idea that there I might miss some of the quadrillion Windows scrap updates (system or "Edge" libraries or whatever). So after installing round about half the size of the internet (and fixing the ~5 FPS in the VM due to ugly transparency effects) I tried to compile the application again, just to see the same result like before. I was close to just postpone the `v2` migration and wait until it is labeled as stable, but then I got one last idea to check if WebView2 is bundled with Windows by default at all. After digging into the [WebView2 documentations][3] I found out that there is indeed a [WebView2 runtime][4] that must be installed individually. Long story short: installing the runtime magically fixed all problems 🎉 

Sorry for the small "Windows rage", but in my opinion the idea to build this OS should have better been discarded back then 🙈
Anyway, is there already a place for `v2` documentations ([wailsapp/docsv2][5]?) except the GitHub repository wiki page? Adding some "Getting Started" information would prevent debugging hours in order to help to test & develop `v2`. The need to explicitly install the WebView2 runtime is an important peace of the puzzle to go through the Windows system architecture hell with a successful application build :smile:
If you are interested in help for a documentation tech-stack I can contribute some recommendations or maybe help to build up the base.

[1]: https://github.com/wailsapp/wails/wiki/Helping-with-V2
[2]: https://github.com/crazy-max/xgo
[3]: https://docs.microsoft.com/en-us/microsoft-edge/webview2
[4]: https://developer.microsoft.com/en-us/microsoft-edge/webview2/#download-section
[5]: https://github.com/wailsapp/docsv2

**Reply by leaanthony** [reply]
I wish I had your patience 😃 when you build for windows it also dumps 2 DLLs with the binary. This is all you should need. If it isn't then i'd like to understand why as those 2 DLLs are literally the webview2 runtime 😃 

Regarding docs, they're a WIP. Expect to hear something shortly!

**Reply by svengreb** [reply]
I also thought that the DLLs are everything that is needed. Maybe it is not possible to simply include them as static files because the runtime differs between Windows update versions. I only know that the installation of the runtime fixed the application loading while the application was stuck when only using the DLLs. Building for macOS works fine since there is no unnecessary clutter needed like DLLs.

I'm currently facing another JS error that prevents the application from loading on the initial start, but works fine when reloading the web view in development mode: `TypeError: undefined is not an object (evaluating 'webkit.messageHandlers.completed.postMessage')`. Is this also one of the functions bound to the `webkit.messageHandlers` object that has not been migrated yet? The error appears right after all initialization events have been fired, but nothing is rendered again because of this error.

**Reply by svengreb** [reply]
I've also noticed that the communication between Go and JS has been slightly changed. The application I currently try to migrate has a function to connect to another (daemon) application via gRPC. The UI has some elements where users can click on a button to connect to the daemon. While the connection is being established the button changes its appearance to show a cross icon and also underlying `onClick` function to cancel the connection again, e.g. when the daemon is not reachable and the user wants to stop the connection process before the timeout is reached.
In Wails `v1` this worked totally fine, but in `v2` the `disconnect` function is not called but only after the timeout has been reached. I've noticed this because I can spam-click on the "cancel" button and nothing gets logged but after the `connect` function returns with a timeout error I one log line per button click. For me this indicates that something changed in the way how the calls are queued. In `v2` function calls to the backend are not really async anymore but somehow blocking.
I've quickly checked the `v2` implementation in #700 and saw that the `n` message type has been removed from the `switch-case` block, but I'm not that familiar with the code yet to see what this type is all about. So in the end I'd like to ask if the new behavior is intended or is this something that needs to be addressed?

**Reply by leaanthony** [reply]
> I also thought that the DLLs are everything that is needed. Maybe it is not possible to simply include them as static files because the runtime differs between Windows update versions. I only know that the installation of the runtime fixed the application loading while the application was stuck when only using the DLLs. Building for macOS works fine since there is no unnecessary clutter needed like DLLs.
> 
> I'm currently facing another JS error that prevents the application from loading on the initial start, but works fine when reloading the web view in development mode: `TypeError: undefined is not an object (evaluating 'webkit.messageHandlers.completed.postMessage')`. Is this also one of the functions bound to the `webkit.messageHandlers` object that has not been migrated yet? The error appears right after all initialization events have been fired, but nothing is rendered again because of this error.

I'll figure out the completed message issue and issue a fix. It must have slipped the net.

> I've also noticed that the communication between Go and JS has been slightly changed. The application I currently try to migrate has a function to connect to another (daemon) application via gRPC. The UI has some elements where users can click on a button to connect to the daemon. While the connection is being established the button changes its appearance to show a cross icon and also underlying `onClick` function to cancel the connection again, e.g. when the daemon is not reachable and the user wants to stop the connection process before the timeout is reached.
> In Wails `v1` this worked totally fine, but in `v2` the `disconnect` function is not called but only after the timeout has been reached. I've noticed this because I can spam-click on the "cancel" button and nothing gets logged but after the `connect` function returns with a timeout error I one log line per button click. For me this indicates that something changed in the way how the calls are queued. In `v2` function calls to the backend are not really async anymore but somehow blocking.
> I've quickly checked the `v2` implementation in #700 and saw that the `n` message type has been removed from the `switch-case` block, but I'm not that familiar with the code yet to see what this type is all about. So in the end I'd like to ask if the new behavior is intended or is this something that needs to be addressed?

Not everything has been implemented yet is the real answer. I'm keen to understand more this scenario though as I haven't hit it.

**Reply by leaanthony** [reply]
@svengreb been looking into this DLL issue. What I can confirm is that simply copying the DLLs doesn't work currently, but I've also discovered that it should: Compiling a test webview application using the same DLLs does indeed work when moving it to a VM without the runtime installed. If I run the example using the same DLLs as Wails generates, it works. The only thing that makes sense is that it's a runtime binding issue - like it can't read those DLLs from the current directory correctly. But then why can the webview example? Will need to step through it with a debugger.

**Reply by leaanthony** [reply]
@svengreb 
> I'm currently facing another JS error that prevents the application from loading on the initial start, but works > fine when reloading the web view in development mode: TypeError: undefined is not an object (evaluating 
> 'webkit.messageHandlers.completed.postMessage').

Is this running using `wails build` or `wails dev`?

**Reply by svengreb** [reply]
> Is this running using wails build or wails dev?

Only when running the desktop application so only when building with the `build` command. In development mode it seem to be not a "critical" error and doesn't prevent the rendering of the application (tested with Firefox, Google Chrome and Safari on macOS). When creating a production build nothing is rendered, either on macOS or Windows.

**Reply by leaanthony** [reply]
I'll test `build` tonight though I'm not sure it's clear how this is happening as this is a native message handler that's nothing to do with the JS runtime!

I've made headway with the DLL issue and can confirm that the issue is partly due to me not fully understanding how the webview2 component works and how to link that against our apps. I've been doing a bit of work on splitting up the code into "webview2" functions and "native wails runtime" functions. I'm not sure when it'll be ready but today's work was a deep understanding of how this all ties together.  For clarity: the `webview.dll` is not needed, at all. Delete it and it'll still work on any machine with the runtime installed. `WebView2Loader.dll` is essentially some code to find where the webview2 library is on the system and load it in at runtime, so it all works. The issue is, how do we get the webview2 runtime on the system? It's a big issue at the moment that's a pain point for a lot of projects. Looking at how the webview project does it, they use an "intermediary" dll which is built using msvc. This allows it to statically link against the webview2 library and pull in what it needs. The resultant DLL is linkable using mingw so that's how we get it to work in Go. The tricky thing is this: Wails generates code for bundling the assets so there needs to be "wails code" that's separate to "webview2" code that gets compiled and linked. I'm in the process of doing that.

**Reply by leaanthony** [reply]
@svengreb I've done some more experiments and got to the bottom of what's going on. It wasn't sitting right with me that webview2 was working for webview apps but not wails apps and that webview.dll was less than a meg, when the webview2 fixed version runtime was 150M uncompressed. And then it hit me: webview drops back to edge rendering engine if it can't load webview2. If you right click a webview app (with debug enabled) on a machine with the webview2 runtime installed, it will work and you'll get a menu and devconsole. If you run the exact same program from a VM with no webview2 runtime, it will fall back to edge - and you can tell by trying to open a dev console: there is none. So, to come full circle, there's nothing I need to do from code perspective to get this working on a fresh machine other than determine what to do with how to get the correct runtime on the user's system. I'll open up a new discussion for that!

**Reply by svengreb** [reply]
On the one hand, that's good news, but on the other, it's not either 😄
So basically the situation for Windows is the same like that of Electron: bring your own runtime or rely on the user to set u the system beforehand. I've saw the new discussion thread in my notifications yesterday and will contribute to it later on.

**Reply by leaanthony** [reply]
It's the same situation as any project using webview2

**Reply by leaanthony** [reply]
@svengreb Got some updates on this on the other discussion - I have a working solution: https://github.com/wailsapp/wails/discussions/736#discussioncomment-870540

**Comment by dwdcth** [comment]
I have tried v2.  It's good enough for me to start a project. Is there any way to add GetPosition and GetWindowSize function to the window api?

**Reply by leaanthony** [reply]
We could look at that. Be aware that it's still alpha so the api may break. What platform are you targeting?

**Reply by dwdcth** [reply]
Ok, I only have Windows now, but my target is crossplatform.
